core/server/master/src/main/java/alluxio/master/AlluxioMasterProcess.java

```diff
@@ -406,7 +418,11 @@ protected void startServingRPCServer() {
   protected void stopServing() throws Exception {
     if (mThriftServer != null) {
       mThriftServer.stop();
+      // Thrift does not interrupt outstanding tasks.
+      // Call shutdownNow() to interrupt them.
+      mExecutorService.shutdownNow();
```

Comment:
@ggezer  Is there any way to detect if all the threads are "done" or not?

Reply:
Can we wait for a period of time to ensure all threads are stopped and the threadpool is no longer accepting tasks? If there are lingering threads still I think it is safe to crash.

Reply:
Thrift already calls shutdown and waits for a configured timeout before returning. Problem is that it doesn't to anything for lingering tasks. So this call will just make sure to terminate terminate them after grace period.

Reply:
If there is a way to detect if threads are still alive, then we should "exit", right? That would ensure that we start the server without any lingering threads.

Reply:
@gpang Are you suggesting a second level of timeout? At the point we called `mExecutor.shutdownNow()` we already waited upto `stopTimeout` for outstanding tasks to finish. `shutdownNow` just interrupts if there any existing threads. After calling `shutdownNow` there is no way to block on completion. We could only see if there are any before invoking it.

Reply:
@ggezer what about `awaitTermination` ? https://docs.oracle.com/javase/7/docs/api/java/util/concurrent/ExecutorService.html#awaitTermination(long,%20java.util.concurrent.TimeUnit)

would we be able to detect with that?